### DISCUSSION TITLE
Is there something similar to ScheduledToastNotification as for UWP apps ?

### DISCUSSION THREAD

**Discussion by Moreno-Borsalino** [microsoft/WindowsAppSDK#3853] [Q&A]
I would like to port a UWP app that uses ScheduledToastNotification but I can't find anything similar in the WinAppSdk notifications documentation.
Can anyone help me and explain if something similar exists or is it a missing feature?

**Comment by castorix** [comment]
It works with _Microsoft.Toolkit.Uwp.Notifications_ package (tested on Windows 10 22H2)
([Schedule a toast notification](https://learn.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/scheduled-toast))

**Comment by Moreno-Borsalino** [comment]
That namespace is only suitable for Windows 10 UWP app project.
But I want to create a WINUI3 desktop application. The documentation says that to use notifications I need to use the "Microsoft.Windows.AppNotifications" namespace. But in that namespace I can't find the ScheduledToastNotifications.
I suspect that they are not supported in the current 1.4 version of WinAppSdk. Or they are hidden somewhere.
Has anyone managed to use them in a WINUI 3 desktop application ?
________________________________
Da: castorix ***@***.***>
Inviato: giovedì 14 settembre 2023 00:38
A: microsoft/WindowsAppSDK ***@***.***>
Cc: Moreno Borsalino ***@***.***>; Author ***@***.***>
Oggetto: Re: [microsoft/WindowsAppSDK] Is there something similar to ScheduledToastNotification as for UWP apps ? (Discussion #3853)


It works with Microsoft.Toolkit.Uwp.Notifications package (tested on Windows 10 22H2)
(Schedule a toast notification<https://learn.microsoft.com/en-us/windows/apps/design/shell/tiles-and-notifications/scheduled-toast>)

**Comment by castorix** [comment]
> That namespace is only suitable for Windows 10 UWP app project.

No, it is for Desktop Apps (with _ToastNotificationManagerCompat_, as the doc says :
_An unpackaged (lacks package identity at runtime) Win32 app must use the ToastNotificationManagerCompat class as seen above. If you use ToastNotificationManager itself, then you'll receive an element-not-found exception. All types of apps can use the Compat class, and it will work correctly._

I tested in WinUI 3/C#, unpackaged Desktop app

**Comment by DarranRowe** [comment]
Toast notifications have always been usable from desktop applications. The issue tends to be toast notifications require a COM server registration. If you look at both WCT's [ToastNotificationManagerCompat](https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/main/Microsoft.Toolkit.Uwp.Notifications/Toasts/Compat/ToastNotificationManagerCompat.cs) and WinAppSDK's [AppNotificationManager](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/AppNotifications/AppNotificationManager.cpp), there is code that registers the application as a COM server. So technically, you don't need the WCT or WinAppSDK to do use a sheduled notification at all. As long as you register things properly, you can just roll your own in any unpackaged desktop application.

The Windows App SDK's App Notifications do not currently support scheduled notifications. However, since App Notifications are just a wrapper around Toast Notifications anyway, try just using ScheduledToastNotification after you register the manager.

**Comment by Moreno-Borsalino** [comment]
I thank @castorix and @DarranRowe for the valuable information that allowed me to understand and solve the problem.
I tried to make a simple example project with a blank Winui3 project. I updated the nuget packages to winappsdk 1.4.
The project creates a window with a button in the center.
I installed the Microsoft.Toolkit.Uwp.Notifications Nuget package.
I added the following lines taken from the Microsoft documentation on Scheduled toast notifications to the code in Click event handler:

// Construct the content and schedule the toast!
new ToastContentBuilder()
    .AddArgument("action", "viewItemsDueToday")
    .AddText("ASTR 170B1")
    .AddText("You have 3 items due today!");
    .Schedule(DateTime.Now.AddSeconds(5));

And with immense joy after 5 seconds the toast notification appears! Even if the application is closed! It's exactly what I was looking for.
